src/dtls/network.c

```diff
@@ -0,0 +1,962 @@
+//
+// The functions below essentially wrap the mbedtls calls adding debugging and
+// handling our data structures.
+//
+
+static void TLSWrite(DPS_NetConnection* cn)
+{
+    int ret;
+    uint8_t* base;
+    DPS_TxBuffer txbuf;
+    size_t total;
+
+    assert(IsNodeThread(cn->node));
+
+    PendingWrite* pw = cn->writeQueue;
+    if (!pw) {
+        DPS_DBGPRINT("TLSWrite() no pending writes\n");
+        return;
+    }
+
+    cn->writeQueue = pw->next;
+    pw->next = NULL;
+
+    {
+        PendingWrite* q = cn->callbackQueue;
+        if (!q) {
+            cn->callbackQueue = pw;
+            uv_idle_start(&cn->idleForSendCallbacks, OnIdleForSendCallbacks);
+        } else {
+            while (q->next) {
+                q = q->next;
+            }
+            q->next = pw;
+        }
+    }
+
+    DPS_DBGPRINT("TLSWrite() using pending write with %d bufs", pw->numBufs);
+
+    if (pw->numBufs == 1) {
+        total = pw->bufs[0].len;
+        base = (uint8_t*)pw->bufs[0].base;
+    } else {
+        total = 0;
+        for (int i = 0; i < pw->numBufs; i++) {
+            total += pw->bufs[i].len;
+        }
+
+        // DPS_NetSend follows libuv and let the user send multiple
+        // buffers. These are automatically merged together. However mbedtls
+        // expects a single buffer.
+        DPS_TxBufferInit(&txbuf, NULL, total);
+        for (int i = 0; i < pw->numBufs; i++) {
+            DPS_TxBufferAppend(&txbuf, (uint8_t*)pw->bufs[i].base, pw->bufs[i].len);
+        }
+        base = txbuf.base;
+    }
+
+    DPS_DBGPRINT("TLSWrite() writing %d bytes of plaintext via DTLS\n", total);
+
+#ifndef NDEBUG
+    for (int i = 0; i < total; i++) {
+        if ((i % 16) == 0) printf("\n"); else printf("  ");
+        printf("%02X", (uint8_t)base[i]);
+    }
+    printf("\n\n");
+#endif
+
+    ret = mbedtls_ssl_write(&cn->ssl, base, total);
+
+    if (pw->numBufs != 1) {
+        DPS_TxBufferFree(&txbuf);
+    }
+
+    // TODO: Need to handle short writes?
+
+    if (ret < 0) {
+        DPS_ERRPRINT("TLSWrite() failure when writing to TLS");
+        pw->status = DPS_ERR_NETWORK;
+    }
+}
+
+static void TLSRead(DPS_NetConnection* cn)
+{
+    DPS_NetContext* netCtx = cn->node->netCtx;
+    int ret;
+
+    assert(IsNodeThread(cn->node));
+
+    ret = mbedtls_ssl_read(&cn->ssl, (unsigned char*)netCtx->plainBuffer, sizeof(netCtx->plainBuffer)-1);
+    if (ret == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+        DPS_DBGPRINT("TLSRead() connection was closed gracefully\n");
+        ret = 0;
+        // TODO: Mark the connection was closed.
+        // TODO: Callback with failure?
+    } else if (ret < 0) {
+        DPS_DBGPRINT("TLSRead() failed, mbedtls_ssl_read returned -0x%x\n\n", -ret);
+        // TODO: Mark the connection was closed.
+        // TODO: Callback with failure?
+    } else {
+        DPS_DBGPRINT("TLSRead() decrypted into %d bytes of plaintext\n", ret);
+
+#ifndef NDEBUG
+        for (int i = 0; i < ret; i++) {
```

Comment:
It's debug code, but keeping since it has been useful when developing DTLS. Fine to remove it later.